src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTableUI.java

```diff
@@ -2076,9 +2082,10 @@ private void paintCells(Graphics g, int rMin, int rMax, int cMin, int cMax) {
         TableColumn aColumn;
         int columnWidth;
         if (table.getComponentOrientation().isLeftToRight()) {
-            for(int row = rMin; row <= rMax; row++) {
+            for (int row = rMin; row <= rMax; row++) {
                 cellRect = table.getCellRect(row, cMin, false);
-                for(int column = cMin; column <= cMax; column++) {
+                cellRect.x = getXPosition(cMin);
```

Comment:
Why do we need to calculate cellRect.x separately? What is wrong in getting the value already there in `cellRect` ?

Reply:
getCellRect spec says
`Returns a rectangle for the cell that lies at the intersection of row and column`

What's the x position offset we are talking here?

Reply:
For RTL orientation [`getCellRect`](https://github.com/openjdk/jdk/blob/6c5e15c1a291ca5ba1e4c3a90351bc71665ce988/src/java.desktop/share/classes/javax/swing/JTable.java#L2977)  returns position excluding the Offset meaning the gap between panel x till x position of table which is x position of first column.  So instead of modifying at JTable class, I am updating here.

Reply:
but we dont seem to have any issue with LTR rendering, so why this getXPosition is needed for LTR too?
Also, wouldn't existing `getRowMargin` have helped?

Reply:
LTR rendering will work unless the Table is fully occupied w.r.t the Panel. Here the Max width of each column is set which is less than the Panel size where extra space would be there on left side of the Table. LTR doesn't need getXPosition because it'll always be 0/attached to Panel at left position. In JTableHeader [r.x](https://github.com/openjdk/jdk/blob/6c5e15c1a291ca5ba1e4c3a90351bc71665ce988/src/java.desktop/share/classes/javax/swing/table/JTableHeader.java#L399) is computed on this offset basis.

Reply:
> LTR doesn't need getXPosition

But it is called in `if (table.getComponentOrientation().isLeftToRight()) `condition above

Reply:
Yes, actually it is not required because it gives/should give the same x position value. Have to remove it, might have added for checking if it had any effect on LTR.

Reply:
I guess you can simplify the fix without using `getXPosition`..You probably would still need `getWidthInRIghtToLeft` but as suggested, having an unified method in a common class like SwingUtilities2 is desirable...Let me know if this will work..

```
--- a/src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTableUI.java
+++ b/src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTableUI.java
@@ -2028,9 +2028,19 @@ private void paintGrid(Graphics g, int rMin, int rMax, int cMin, int cMax) {
         if (table.getShowHorizontalLines()) {
             int tableWidth = damagedArea.x + damagedArea.width;
             int y = damagedArea.y;
-            for (int row = rMin; row <= rMax; row++) {
-                y += table.getRowHeight(row);
-                SwingUtilities2.drawHLine(g, damagedArea.x, tableWidth - 1, y - 1);
+            if (table.getComponentOrientation().isLeftToRight()) {
+                for (int row = rMin; row <= rMax; row++) {
+                    y += table.getRowHeight(row);
+                    SwingUtilities2.drawHLine(g, damagedArea.x, tableWidth - 1, y - 1);
+                }
+            } else {
+                TableColumnModel cm = table.getColumnModel();
+                for (int row = rMax; row >= rMin; row--) {
+                    y += table.getRowHeight(row);
+                    int w = cm.getColumn(cMax).getWidth();
+                    damagedArea.x = table.getWidth() - (w * (cMax + 1));
+                    SwingUtilities2.drawHLine(g, damagedArea.x, table.getWidth() - 1, y - 1);
+                }
             }
         }
         if (table.getShowVerticalLines()) {
@@ -2090,18 +2100,12 @@ private void paintCells(Graphics g, int rMin, int rMax, int cMin, int cMax) {
             }
         } else {
             for(int row = rMin; row <= rMax; row++) {
-                cellRect = table.getCellRect(row, cMin, false);
-                aColumn = cm.getColumn(cMin);
-                if (aColumn != draggedColumn) {
-                    columnWidth = aColumn.getWidth();
-                    cellRect.width = columnWidth - columnMargin;
-                    paintCell(g, cellRect, row, cMin);
-                }
-                for(int column = cMin+1; column <= cMax; column++) {
+                for(int column = cMin; column <= cMax; column++) {
+                    cellRect = table.getCellRect(row, column, false);
                     aColumn = cm.getColumn(column);
                     columnWidth = aColumn.getWidth();
                     cellRect.width = columnWidth - columnMargin;
-                    cellRect.x -= columnWidth;
+                    cellRect.x = table.getWidth() - (columnWidth * (column + 1));
                     if (aColumn != draggedColumn) {
                         paintCell(g, cellRect, row, column);
```

Reply:
@prsadhuk Its working if the width of all columns are equal, and doesn't work if we have varying column width.